lib/rubygems.rb

```diff
@@ -1293,7 +1293,12 @@ def already_loaded?(file)
     end
 
     def default_gem_load_paths
-      @default_gem_load_paths ||= $LOAD_PATH[load_path_insert_index..-1]
+      @default_gem_load_paths ||= $LOAD_PATH[load_path_insert_index..-1].map do |lp|
+        expanded = File.expand_path(lp)
+        next expanded unless File.exist?(expanded)
+
+        File.realpath(expanded)
```

Comment:
I am afraid that using `require_relative` is digging its own rabbit hole and solves no issue. I wish Rubygems returned to using plain require. There are at least two reasons:

1) `require_relative` does not avoid double loading of files, it just shifts the problem elsewhere
2) The real issue is actually class reopening and redefinition attempts, which it does not avoid either.

Reply:
Although in Ruby 3.1, there is ruby/ruby/pull/4887, so maybe the situation might be a tiny bit better.

Reply:
I don't use `require_relative` for any of those reasons. I use `require_relative` because it's faster and simpler to understand for requiring internal code. We are not going back to plain `require` since everybody agrees here, I've been proposing PRs in this repository and many others and you have been the only person against it in all this time.

Reply:
By the way, maybe this diff is confusing because I changed some `autoload`s to `require_relative` while investigating, but `require_relative` is completely unrelated to this issue. Or at least I don't get how it's connected.

Reply:
I don't see why are you suddenly need to expand the paths. That was not needed before and the only reason I can imagine is the `require_relative` and its incompatibility with `require`. The commit message does not provide more insights. Maybe it is completely unrelated.

Reply:
Yeah, it's completely unrelated, I'll improve the commit message.